pkg/route/secret/secret_monitor.go

```diff
@@ -0,0 +1,217 @@
+		if !cache.WaitForCacheSync(ctx.Done(), m.itemMonitor.HasSynced) {
+			return nil, fmt.Errorf("failed waiting for cache sync")
+		}
+
+		// add item key to monitors map // add watch to the list
+		s.monitors[key] = m
+
+		klog.Info("secret informer started", " item key ", key)
+	}
+
+	// add the event handler
+	registration, err := m.itemMonitor.AddEventHandler(handler)
+	if err != nil {
+		return nil, err
+	}
+
+	// Increment numHandlers
+	m.numHandlers += 1
+
+	// TODO: this can be too noisy, later we need to use higher verbosity
+	klog.Info("secret handler added", " item key ", key)
+
+	return registration, nil
+}
+
+// RemoveSecretEventHandler removes a secret event handler and stops the informer if no handlers are left.
+// If the handler is not found or if there is an issue removing it, an error is returned.
+func (s *secretMonitor) RemoveSecretEventHandler(handlerRegistration SecretEventHandlerRegistration) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if handlerRegistration == nil {
+		return fmt.Errorf("nil secret handler registration is provided")
+	}
+
+	// Extract the key from the registration to identify the associated monitor.
+	// populated in AddEventHandler()
+	key := handlerRegistration.GetKey()
+
+	// check if secret informer already exists for the secret(key)
+	m, exists := s.monitors[key]
+	if !exists {
+		return fmt.Errorf("secret monitor already removed for item key %v", key)
+	}
+
+	if err := m.itemMonitor.RemoveEventHandler(handlerRegistration); err != nil {
+		return err
+	}
+	// Decrement numHandlers
+	m.numHandlers -= 1
+	klog.Info("secret handler removed", " item key", key)
+
+	// stop informer if there is no handler
+	if m.numHandlers <= 0 {
+		if !m.itemMonitor.StopInformer() {
+			return fmt.Errorf("secret informer already stopped for item key %v", key)
+		}
+		// remove the key from map
+		delete(s.monitors, key)
+		klog.Info("secret informer stopped", " item key ", key)
+	}
+
+	return nil
+}
+
+// GetSecret retrieves the secret object from the informer's cache. Error if the secret is not found in the cache.
+func (s *secretMonitor) GetSecret(ctx context.Context, handlerRegistration SecretEventHandlerRegistration) (*corev1.Secret, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if handlerRegistration == nil {
+		return nil, fmt.Errorf("nil secret handler registration is provided")
+	}
+	key := handlerRegistration.GetKey()
+	secretName := key.Name
+
+	// check if secret informer exists
+	m, exists := s.monitors[key]
+	if !exists {
+		return nil, fmt.Errorf("secret monitor doesn't exist for key %v", key)
+	}
+
+	// wait for informer store sync, to load secrets
+	if !cache.WaitForCacheSync(ctx.Done(), handlerRegistration.HasSynced) {
+		return nil, fmt.Errorf("failed waiting for cache sync")
+	}
+
+	uncast, exists, err := m.itemMonitor.GetItem()
+	if !exists {
+		return nil, apierrors.NewNotFound(corev1.Resource("secrets"), secretName)
+	}
+
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Informer's cache (`GetByKey()`) is returning `error` even with `exists` , so we should handle `error` instead of ignoring it.

Reply:
Sorry, I didn't make the question clear. I meant that why not checking the error before we check exist. It may be misleading as `exist` may be false just because it's a default value returned in case of an error. This may result in a wrong error.

Reply:
Okay, that makes sense. Moved `err` check before `exist`. Thanks!